tests/bosonic_files/test_backend_bosoniccircuit.py

```diff
@@ -0,0 +1,427 @@
+        assert example.get_mean().shape == (tot_num_weights, num_quad)
+        assert example.get_covmat().shape == circuit.c_shape(num_modes, num_weights)
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("num_modes", NUM_MODES_VALS)
+    def test_reset_circuit(self, num_weights, num_modes):
+        r"""Checks that the reset method instantiates the correct number of modes
+        as vacuum, each with a given number of weights."""
+        # Create a state over 1 mode, 1 weight per mode
+        example = circuit.BosonicModes(1, 1)
+
+        # Reset with number of weights and modes and perform the same check
+        # as test_circuit_init
+        example.reset(num_modes, num_weights)
+        tot_num_weights = num_weights ** num_modes
+        num_quad = 2 * num_modes
+        assert example.nlen == num_modes
+        assert np.isclose(sum(example.get_weights()), 1)
+        assert example.is_vacuum(tol=1e-10)
+        for i in range(num_modes):
+            assert np.isclose(example.fidelity_vacuum([i]), 1)
+        assert example.active == list(range(num_modes))
+        assert example.get_mean().shape == (tot_num_weights, num_quad)
+        assert example.get_covmat().shape == circuit.c_shape(num_modes, num_weights)
+        to_xp_list = list(range(0, num_quad, 2)) + list(range(1, num_quad + 1, 2))
+        from_xp_list = []
+        for i in range(num_modes):
+            from_xp_list += [i, i + num_modes]
+        assert np.allclose(example.to_xp, np.array(to_xp_list))
+        assert np.allclose(example.from_xp, np.array(from_xp_list))
+
+    @pytest.mark.parametrize("num_weights", NUM_WEIGHTS_VALS)
+    @pytest.mark.parametrize("num_modes", NUM_MODES_VALS)
+    @pytest.mark.parametrize("num_weights_new", NUM_NEW_WEIGHTS_VALS)
+    def test_add_del_modes(self, num_weights, num_modes, num_weights_new):
+        r"""Checks that modes can be added and deleted."""
+        # Create a state with the pre-defined number of weights and modes
+        example = circuit.BosonicModes(num_modes, num_weights)
+        # Delete first mode and check active modes changed
+        example.del_mode(0)
+        assert example.get_modes() == list(range(1, num_modes))
+
+        # Add new mode with new number of weights
+        example.add_mode([num_weights_new])
+        tot_weights = (num_weights ** num_modes) * num_weights_new
+        # Check numbers of weights, means and covs
+        assert example.get_modes() == list(range(1, num_modes + 1))
+        assert np.isclose(sum(example.get_weights()), 1)
+        assert example.get_weights().shape == (tot_weights,)
+        assert example.get_mean().shape == (tot_weights, 2 * (num_modes + 1))
+        assert example.get_covmat().shape == (tot_weights, 2 * (num_modes + 1), 2 * (num_modes + 1))
+
+    @pytest.mark.parametrize("phi", PHI_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_displace_squeeze_rotate(self, r, phi):
```

Comment:
This case is testing too much, e.g., squeezing and applying a phase shift could be tested separately.

Reply:
Done.